pandas/io/excel/_base.py

```diff
@@ -0,0 +1,852 @@
+            self.datetime_format = 'YYYY-MM-DD HH:MM:SS'
+        else:
+            self.datetime_format = datetime_format
+
+        self.mode = mode
+
+    def __fspath__(self):
+        return _stringify_path(self.path)
+
+    def _get_sheet_name(self, sheet_name):
+        if sheet_name is None:
+            sheet_name = self.cur_sheet
+        if sheet_name is None:  # pragma: no cover
+            raise ValueError('Must pass explicit sheet_name or set '
+                             'cur_sheet property')
+        return sheet_name
+
+    def _value_with_fmt(self, val):
+        """Convert numpy types to Python types for the Excel writers.
+
+        Parameters
+        ----------
+        val : object
+            Value to be written into cells
+
+        Returns
+        -------
+        Tuple with the first element being the converted value and the second
+            being an optional format
+        """
+        fmt = None
+
+        if is_integer(val):
+            val = int(val)
+        elif is_float(val):
+            val = float(val)
+        elif is_bool(val):
+            val = bool(val)
+        elif isinstance(val, datetime):
+            fmt = self.datetime_format
+        elif isinstance(val, date):
+            fmt = self.date_format
+        elif isinstance(val, timedelta):
+            val = val.total_seconds() / float(86400)
+            fmt = '0'
+        else:
+            val = compat.to_str(val)
+
+        return val, fmt
+
+    @classmethod
+    def check_extension(cls, ext):
+        """checks that path's extension against the Writer's supported
+        extensions.  If it isn't supported, raises UnsupportedFiletypeError."""
+        if ext.startswith('.'):
+            ext = ext[1:]
+        if not any(ext in extension for extension in cls.supported_extensions):
+            msg = (u("Invalid extension for engine '{engine}': '{ext}'")
+                   .format(engine=pprint_thing(cls.engine),
+                           ext=pprint_thing(ext)))
+            raise ValueError(msg)
+        else:
+            return True
+
+    # Allow use as a contextmanager
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def close(self):
+        """synonym for save, to make it more file-like"""
+        return self.save()
+
+
+class ExcelFile(object):
+    """
+    Class for parsing tabular excel sheets into DataFrame objects.
+    Uses xlrd. See read_excel for more documentation
+
+    Parameters
+    ----------
+    io : string, path object (pathlib.Path or py._path.local.LocalPath),
+        file-like object or xlrd workbook
+        If a string or path object, expected to be a path to xls or xlsx file.
+    engine : string, default None
+        If io is not a buffer or path, this must be set to identify io.
+        Acceptable values are None or ``xlrd``.
+    """
+
+    from pandas.io.excel._xlrd import _XlrdReader
```

Comment:
Had to move this into the class as it would be a circular reference otherwise, though ideally this shouldn't be required at all in the base class